nntrainer/layers/plugged_layer.h

```diff
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   plugged_layer.h
+ * @date   27 January 2021
+ * @brief  This file contains a wrapper for a plugged layer, INTERNAL USE ONLY
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#ifndef __PLUGGED_LAYER_H__
+#define __PLUGGED_LAYER_H__
+
+#include <layer.h>
+
+#include <layer_internal.h>
+#include <manager.h>
+#include <nntrainer_error.h>
+
+namespace nntrainer {
+namespace internal {
+
+/**
+ * @brief PluggedLayer to wrap a layer from shared object file
+ *
+ */
+class PluggedLayer : public nntrainer::Layer {
+public:
+  /**
+   * @brief Construct a new Plugged Layer object
+   *
+   * @param pluggable LayerPluggable structure from the symbol
+   */
+  PluggedLayer(const ml::train::LayerPluggable *pluggable) :
+    /// @todo we won't need dynamic pointer cast here after api is fully
+    /// implemented
+    layerImpl(dynamic_cast<nntrainer::Layer *>(pluggable->createfunc())),
+    destroy_func(pluggable->destroyfunc) {
+    NNTR_THROW_IF(layerImpl == nullptr, std::invalid_argument)
+      << "either create_func_ failed or cannot dynamic cast to layer_internal";
+  }
+
+  /**
+   * @brief Destroy the Plugged Layer object
+   *
+   */
+  ~PluggedLayer() override { destroy_func(layerImpl); }
+
+  /**
+   * @brief Move Contruct Plugged Layer object
+   *
+   * @param rhs layer to move
+   */
+  PluggedLayer(PluggedLayer &&rhs) noexcept = default;
+
+  /**
+   * @brief Move assign Plugged Layer Object
+   *
+   * @param rhs layer to move
+   * @return PluggedLayer& *this
+   */
+  PluggedLayer &operator=(PluggedLayer &&rhs) = default;
+
+  /**
+   * @copydoc Layer::initialize(Manager &manager)
+   */
+  int initialize(Manager &manager) { return layerImpl->initialize(manager); }
+
+  /**
+   * @copydoc Layer::forwarding(bool training)
+   */
+  void forwarding(bool training = true) override {
+    layerImpl->forwarding(training);
+  }
+
+  /**
+   * @copydoc Layer::calcDerivative()
+   */
+  void calcDerivative() override { layerImpl->calcDerivative(); }
+
+  /**
+   * @copydoc Layer::calcGradient()
+   */
+  void calcGradient() override { layerImpl->calcGradient(); }
+
+  /**
```

Comment:
don't we need layerImpl->applyGradient ??

Reply:
Yes, thank you for noticing hardly noticeable issue :)

Reply:
Updated